src/org/daisy/dotify/formatter/impl/core/TableOfContentsImpl.java

```diff
@@ -17,40 +33,188 @@
+		}
+		currentAncestorTocBlocks.push(tocBlock);
+		super.startBlock(p, blockId);
 	}
-	
-	public Set<String> getTocIdList() {
-		return refs.keySet();
+
+	@Override
+	public void endBlock() {
+		super.endBlock();
+		currentAncestorTocBlocks.pop();
 	}
-	
-	public String getRefForID(String id) {
-		return refs.get(id);
+
+	@Override
+	public Block newBlock(String blockId, RowDataProperties rdp) {
+		Block b = super.newBlock(blockId, rdp);
+		if (!currentAncestorTocBlocks.isEmpty()) {
+			tocBlockForBlock.put(b, currentAncestorTocBlocks.peek());
+		}
+		return b;
 	}
 
 	@Override
-	public void startEntry(String refId, BlockProperties props) {
-		String tocId;
-		do {
-			tocId = ""+((int)Math.round(99999999*Math.random()));
-		} while (containsTocID(tocId));
-		if (refs.put(tocId, refId)!=null) {
-			throw new RuntimeException("Identifier is not unique: " + tocId);
+	public void startEntry(String refId) {
+		if (inTocEntry) {
+			throw new RuntimeException("toc-entry may not be nested");
+		}
+		inTocEntry = true;
+		if (!refIds.add(refId)) {
+			throw new RuntimeException("ref-id is not unique: " + refId);
+		}
+		if (refIdForBlock.put(getCurrentBlock(), refId) != null) {
+			// note that this is not strictly forbidden by OBFL, but it simplifies the implementation
+			throw new RuntimeException("No two toc-entry's may be contained in the same block");
 		}
-		startBlock(props, tocId);
 	}
 	
 	@Override
 	public void endEntry() {
-		endBlock();
+		inTocEntry = false;
 	}
 
+	/**
+	 * Filter out the toc-entry with a ref-id that does not satisfy the predicate. This is used to
+	 * create the volume range toc. toc-block that have all their descendant toc-entry filtered out
+	 * are also omitted.
+	 */
+	/*
+	 * Note that, because this is implemented by filtering a fixed sequence of blocks, and because
+	 * of the way the sequence of blocks is constructed, we are potentially throwing away borders
+	 * and margins that should be kept. That said, the previous implementation did not handle
+	 * borders and margins correctly either, so fixing this issue can be seen as an optimization.
+	 */
+	public Collection<Block> filter(Predicate<String> filter) {
+		List<Block> filtered = new ArrayList<>();
+		Set<Object> tocBlocksWithDescendantTocEntry = new HashSet<>();
+		for (Block b : this) {
+			if (refIdForBlock.containsKey(b)) {
+				if (!filter.test(refIdForBlock.get(b))) {
+					continue;
+				}
+				if (tocBlockForBlock.containsKey(b)) {
+					Object tocBlock = tocBlockForBlock.get(b);
+					tocBlocksWithDescendantTocEntry.add(tocBlock);
+					while (parentTocBlockForTocBlock.containsKey(tocBlock)) {
+						tocBlock = parentTocBlockForTocBlock.get(tocBlock);
+						tocBlocksWithDescendantTocEntry.add(tocBlock);
+					}
+				}
+			}
+			filtered.add(b);
+		}
+		Iterator<Block> i = filtered.iterator();
+		while (i.hasNext()) {
+			Block b = i.next();
+			if (refIdForBlock.containsKey(b)) {
```

Comment:
I'm not sure about this - but I would expect here a `if (!refIdForBlock.containsKey(b))`

Reply:
It's correct alright. In this loop we remove the `Block`s that came from a `toc-block` that don't contain any rendered `toc-entry`. `refIdForBlock.containsKey(b)` means the `Block` contains a `toc-entry`, and the `continue` on the next line means keep the block.

Reply:
It would be nice to have a unit test for the part of the code that removes a `toc-block` without descendant rendered `toc-entry`s.

Reply:
Yes good idea.